common/lib/xmodule/xmodule/library_content_module.py

```diff
@@ -147,6 +204,56 @@ def _publish_event(self, event_name, result, **kwargs):
         self.runtime.publish(self, "edx.librarycontentblock.content.{}".format(event_name), event_data)
         self._last_event_result_count = len(result)  # pylint: disable=attribute-defined-outside-init
 
+    @classmethod
+    def publish_selected_children_events(cls, block_keys, format_block_keys, publish_event):
+        """
+        Publish selected events on child objects.  # REVIEW: Useless. Rewrite.
+        Dynamically selects block_ids indicating which of the possible children are displayed to the current user.
+
+        Arguments:
+
+            block_keys -
+                A dict describing which events to publish (add or remove), see `make_selection` above
+                for format details.
+
+            format_block_keys -
+                A function to convert block keys to the format expected by publish_event. Must have the signature:
+
+                    [(block_type, block_id)] -> T
+
+                Where T is a collection of block keys as accepted by `publish_event`.
+
+            publish_event -
+                Function that handles the actual publishing.  Must have the signature:
+
+                    <'removed'|'assigned'> -> result:T -> removed:T -> reason:basestring -> None
+
+                Where T is a collection of block_keys as returned by `format_block_keys`.
+        """
+        if block_keys['invalid']:
+            # reason "invalid" means deleted from library or a different library is now being used.
+            publish_event(
+                "removed",
+                result=format_block_keys(block_keys['selected']),
+                removed=format_block_keys(block_keys['invalid']),
+                reason="invalid"
+            )
+
+        if block_keys['overlimit']:
+            publish_event(
+                "removed",
+                result=format_block_keys(block_keys['selected']),
+                removed=format_block_keys(block_keys['overlimit']),
+                reason="overlimit"
+            )
+
+        if block_keys['added']:
+            publish_event(
+                "assigned",
+                result=format_block_keys(block_keys['selected']),
+                added=format_block_keys(block_keys['added'])
```

Comment:
I was just thinking all this could be factored:

```
cases = [
    ('invalid', 'removed', 'invalid'),
    ('overlimit', 'removed', overlimit'),
    ('added', 'assigned', None),
]
for (event, key, reason) in cases:
 ... stuff ...
```

But I'm not sure if passing ``reason=None`` is the right thing (though it would be easy enough to filter that out).